### PR TITLE
UNR-481: Pending Replicated Property Updates Resolve Correctly to Owning Object

### PR DESCRIPTION
**Pull request by wanqizhu**
#### Description
See https://improbableio.atlassian.net/browse/UNR-481

Tldr: When there are pending object refs for property updates, e.g. when you cross over a worker boundary and your PlayerController, which has a pointer to PlayerState, is spawned before PlayerState, we need to correctly apply replicated properties inside PlayerState, link it up with PlayerController, and notify PlayerController's OnRepNotifies. This is currently broken, especially when we resolve a property owned not by the root actor but by a sub ActorComponent.

We change typebinding and how we store PendingIncomingObjectUpdates to keep track of the owning object (TargetObject, either Actor or ActorComponent but not necessarily the owning actor of the root ActorChannel). This way, when we resolve our IncomingObject, we correctly link it with TargetObject->UnresolvedObjectRef as a property field and call corresponding OnRepNotifies.

**Note that typebinding will need to be regenerated.**

#### Tests
Running across worker boundaries no longer throw warning: "can't find onrep-notify function." Inspecting shows that indeed, the PlayerController's Pawn starts out NULL and gets set to the correct IncomingObject.

Need to explicitly test for incoming property updates for subobjects.

#### Primary reviewers
@improbable-valentyn @m-samiec

### CODE REVIEW

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 1]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
 		USpatialActorChannel* DependentChannel = ChannelProperties.Key;
 		FPendingIncomingProperties& Properties = ChannelProperties.Value;
 
-		// Trigger pending updates.
-		DependentChannel->PreReceiveSpatialUpdate(Object);
-
+		// Replicated Properties
 		TSet<UProperty*> RepNotifies;
-		for (const FRepHandleData* RepData : Properties.Key)
+		for (auto& ObjsToReplicate : Properties.Key)  // TPair<UObject*, TArray<const FRepHandleData*>>
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/SpatialInterop.cpp]: No point using auto if we've the type commented there also. Remove the comment, or replace auto with the type.

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 2]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
...
-				*SpatialOSInstance->GetWorkerId(),
-				*DependentChannel->Actor->GetName(),
-				DependentChannel->GetEntityId().ToSpatialEntityId(),
-				*RepData->Property->GetName());
+			// Iterate through our list of target objects, either an actor or component
+			// For each, we resolve all replicated properties under that object
+			UObject* TargetObject = ObjsToReplicate.Key;
+			DependentChannel->PreReceiveSpatialUpdate(TargetObject);
+
+			for (const FRepHandleData* RepData : ObjsToReplicate.Value) {  // Iterating through TArray<const FRepHandleData*>
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/SpatialInterop.cpp]: Unnecessary comment. Generally comment why, not what.

File: Source/SpatialGDK/Public/SpatialInterop.h [hunk 3]
```diff
@@ -93,10 +94,12 @@ using FOutgoingPendingArrayUpdateMap = TMap<const UObject*, FChannelToHandleToOP
 
 using FResolvedObjects = TArray<TPair<UObject*, const improbable::unreal::UnrealObjectRef>>;
 
-// Helper function to write incoming replicated property data to an object.
-FORCEINLINE void ApplyIncomingReplicatedPropertyUpdate(const FRepHandleData& RepHandleData, UObject* Object, const void* Value, TSet<UProperty*>& RepNotifies)
+// Helper function to write incoming replicated property data to an object
+// TargetObject is the actor or actor component that owns the replicated property and its RepNotifies, NOT the root actor channel
+// to-do: unify how replicated properties and handover are handled
```
Comment (m-samiec) [inline Source/SpatialGDK/Public/SpatialInterop.h]: I'm confused, isn't this function dealing with them in the same way? How do you want to unify them more? 
And if you add a todo, please add a tracking Jira ticket.

Comment (m-samiec) [reply Source/SpatialGDK/Public/SpatialInterop.h]: Ahh don't mind me, I see they have different property updaters :)
Please add a Jira ticket though.

File: Source/SpatialGDK/Public/SpatialInterop.h [hunk 4]
```diff
@@ -60,11 +60,12 @@ class FOutgoingReliableRPC
 };
 
 // Helper types used by the maps below.
-using FPendingOutgoingProperties = TPair<TArray<uint16>, TArray<uint16>>; // Pending incoming properties (replicated and handover).
-using FPendingIncomingProperties = TPair<TArray<const FRepHandleData*>, TArray<const FHandoverHandleData*>>;
+using FPendingOutgoingProperties = TPair<TArray<uint16>, TArray<uint16>>;
+using FPendingIncomingProperties = TPair<TMap<UObject*, TArray<const FRepHandleData*>>, TArray<const FHandoverHandleData*>>;  // Pending incoming properties (replicated and handover).
```
Comment (m-samiec) [inline Source/SpatialGDK/Public/SpatialInterop.h]: This is a dense typedef, and though to parse. Worth a comment explaining the relationships between the types.
Or breaking the first map into it's own typedef.

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 5]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
...
+				ApplyIncomingReplicatedPropertyUpdate(*RepData, TargetObject, &IncomingObject, RepNotifies);
+				UE_LOG(LogSpatialGDKInterop, Log, TEXT("%s: Received queued object replicated property update. Channel actor (root actor) %s (%lld), target object %s, property %s"),
```
Comment (improbable-valentyn) [inline Source/SpatialGDK/Private/SpatialInterop.cpp]: I don't think "root actor" is a thing. I feel like "Channel actor" is descriptive enough.

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 6]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
...
+
+			DependentChannel->PostReceiveSpatialUpdate(TargetObject, RepNotifies.Array());
 		}
+
+		// Handover
 		for (const FHandoverHandleData* HandoverData : Properties.Value)
 		{
-			ApplyIncomingHandoverPropertyUpdate(*HandoverData, DependentChannel->Actor, &Object);
-			UE_LOG(LogSpatialGDKInterop, Log, TEXT("%s: Received queued object handover property update. actor %s (%lld), property %s"),
+			ApplyIncomingHandoverPropertyUpdate(*HandoverData, DependentChannel->Actor, &IncomingObject);  // HandoverHandleData will find the correct TargetObject for our replicated property, so we can just pass in the root actor
```
Comment (improbable-valentyn) [inline Source/SpatialGDK/Private/SpatialInterop.cpp]: Channel actor, or owning actor instead of root actor.

File: Source/SpatialGDK/Public/SpatialInterop.h [hunk 7]
```diff
@@ -93,10 +94,12 @@ using FOutgoingPendingArrayUpdateMap = TMap<const UObject*, FChannelToHandleToOP
 
 using FResolvedObjects = TArray<TPair<UObject*, const improbable::unreal::UnrealObjectRef>>;
 
-// Helper function to write incoming replicated property data to an object.
-FORCEINLINE void ApplyIncomingReplicatedPropertyUpdate(const FRepHandleData& RepHandleData, UObject* Object, const void* Value, TSet<UProperty*>& RepNotifies)
+// Helper function to write incoming replicated property data to an object
+// TargetObject is the actor or actor component that owns the replicated property and its RepNotifies, NOT the root actor channel
```
Comment (improbable-valentyn) [inline Source/SpatialGDK/Public/SpatialInterop.h]: "NOT the root actor channel" is misleading. If the property is on the actor, the `TargetObject` will in fact be the channel actor. I think the first part already explains it well, so just remove the "NOT the root actor channel" part.

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 1 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: done

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 2 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: done

File: Source/SpatialGDK/Public/SpatialInterop.h (hunk 3 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Public/SpatialInterop.h]: done

File: Source/SpatialGDK/Public/SpatialInterop.h (hunk 4 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Public/SpatialInterop.h]: done

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 5 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: done

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 6 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: done

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 8]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
...
-				*SpatialOSInstance->GetWorkerId(),
-				*DependentChannel->Actor->GetName(),
-				DependentChannel->GetEntityId().ToSpatialEntityId(),
-				*RepData->Property->GetName());
+			// Iterate through our list of target objects, either an actor or component
+			// For each, we resolve all replicated properties under that object
+			UObject* TargetObject = ObjsToReplicate.Key;
+			DependentChannel->PreReceiveSpatialUpdate(TargetObject);
+
+			for (const FRepHandleData* RepData : ObjsToReplicate.Value) {
```
Comment (improbable-valentyn) [inline Source/SpatialGDK/Private/SpatialInterop.cpp]: Nit: `{` should be on a new line.

File: Source/SpatialGDK/Public/SpatialInterop.h [hunk 9]
```diff
@@ -93,10 +95,12 @@ using FOutgoingPendingArrayUpdateMap = TMap<const UObject*, FChannelToHandleToOP
 
 using FResolvedObjects = TArray<TPair<UObject*, const improbable::unreal::UnrealObjectRef>>;
 
-// Helper function to write incoming replicated property data to an object.
-FORCEINLINE void ApplyIncomingReplicatedPropertyUpdate(const FRepHandleData& RepHandleData, UObject* Object, const void* Value, TSet<UProperty*>& RepNotifies)
+// Helper function to write incoming replicated property data to an object
+// TargetObject is the actor or actor component that owns the replicated property and its RepNotifies, not necessarily the channel actor
+// to-do: unify how replicated properties and handover are handled (UNR-485)
```
Comment (improbable-valentyn) [inline Source/SpatialGDK/Public/SpatialInterop.h]: Nit: `TODO: `

File: Source/SpatialGDK/Private/SpatialInterop.cpp [hunk 10]
```diff
@@ -613,30 +613,38 @@ void USpatialInterop::ResolvePendingIncomingObjectUpdates(UObject* Object, const
 		USpatialActorChannel* DependentChannel = ChannelProperties.Key;
 		FPendingIncomingProperties& Properties = ChannelProperties.Value;
 
-		// Trigger pending updates.
-		DependentChannel->PreReceiveSpatialUpdate(Object);
-
+		// Replicated Properties
 		TSet<UProperty*> RepNotifies;
-		for (const FRepHandleData* RepData : Properties.Key)
+		for (TPair<UObject*, TArray<const FRepHandleData*>>& ObjsToReplicate : Properties.Key)
```
[the remark this answers is not in the document]
Comment (improbable-valentyn) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: The map is typedef'd, but the pair is just the type used to iterate the map. I personally don't feel the need for a typedef, but if you feel strongly about it I don't mind.

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 8 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: done

File: Source/SpatialGDK/Public/SpatialInterop.h (hunk 9 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Public/SpatialInterop.h]: done

File: Source/SpatialGDK/Private/SpatialInterop.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: yeah agree with Valentyn -- it's only used here instead of auto& for verbosity

[the remark this answers is not in the document]
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInterop.cpp]: agree with naming change
